apps-rendering/src/components/TableOfContent/index.tsx

```diff
@@ -0,0 +1,193 @@
+	}
+`;
+
+const listItemStyles: SerializedStyles = css`
+	${textSans.xsmall({ fontWeight: 'bold', lineHeight: 'regular' })}
+	border-bottom: ${line.primary} 1px solid;
+	padding-top: ${remSpace[2]};
+`;
+
+const detailsStyles: SerializedStyles = css`
+	margin-bottom: ${remSpace[6]};
+	&:not([open]) .is-on,
+	&[open] .is-off {
+		display: none;
+	}
+	summary::-webkit-details-marker {
+		display: none;
+	}
+`;
+
+const summaryStyles: SerializedStyles = css`
+	cursor: pointer;
+	position: relative;
+	list-style: none;
+	align-items: center;
+	padding-left: ${remSpace[3]};
+	padding-top: 0.44rem;
+	padding-bottom: 0.375rem;
+	border-bottom: ${line.primary} 1px solid;
+	border-top: ${line.primary} 1px solid;
+
+	path {
+		fill: ${neutral[46]};
+	}
+	svg {
+		height: 2rem;
+	}
+`;
+
+const titleStyle = (format: ArticleFormat): SerializedStyles => css`
+	${textSans.xsmall({ lineHeight: 'regular' })}
+	color: ${text.tableOfContentsTitle(format)};
+	${darkModeCss`
+		color: ${text.tableOfContentsTitleDark(format)};
+	`}
+`;
+
+const arrowPosition: SerializedStyles = css`
+	position: absolute;
+	right: ${remSpace[1]};
+	top: 0;
+`;
+
+const TocTextElement: React.FC<TextElementProps> = ({
+	node,
+	key,
+}): ReactElement => {
+	const text = node.textContent ?? '';
+	const children = Array.from(node.childNodes).map((item, i) => {
+		console.log(`I'm children node ${item.nodeName}`);
+		return <TocTextElement node={item} key={i} />;
+	});
+
+	console.log(`I'm node ${node.nodeName}`);
+
+	switch (node.nodeName) {
+		case 'H2':
+			return <span>{children}</span>;
+		case 'EM':
+			return <em key={key}>{children}</em>;
+		case 'SUB': {
+			return (
+				<sub
+					css={css`
+						font-size: smaller;
+						vertical-align: sub;
+					`}
+					key={key}
+				>
+					{children}
+				</sub>
+			);
+		}
+		case 'SUP': {
+			return (
+				<sup
+					css={css`
+						font-size: smaller;
+						vertical-align: super;
+					`}
+					key={key}
+				>
+					{children}
+				</sup>
+			);
+		}
+		case 'STRONG':
+			return (
+				<strong
+					css={css`
+						font-weight: bold;
+					`}
+					key={key}
+				>
+					{children}
+				</strong>
+			);
+		case '#text':
+		default:
+			return <>{text}</>;
+	}
+};
+
+const TableOfContent: FC<Props> = ({ format, outline }) => {
+	return (
+		<details open={outline.length < 5} css={detailsStyles}>
+			<summary css={summaryStyles}>
+				<h2 css={titleStyle(format)}>Jump to...</h2>
+				<span className="is-off" css={arrowPosition}>
+					<SvgChevronDownSingle size="xsmall" />
+				</span>
+				<span className="is-on" css={arrowPosition}>
+					<SvgChevronUpSingle size="xsmall" />
+				</span>
+			</summary>
+			<OrderedList className={listStyles}>
+				{outline.map((ol) => (
+					<ListItem className={listItemStyles}>
+						<Anchor
+							format={format}
+							href={`#${ol.id}`}
+							className={anchorStyles(format)}
+						>
+							<TocTextElement
+								node={ol.doc}
+								key={0}
+							></TocTextElement>
```

Comment:
Thanks, by the way, do you think using 0 for key is OK? I'm so confused about this key thing in react 🤷

Reply:
This is a technical look at why keys are important in React: https://reactjs.org/docs/reconciliation.html#recursing-on-children

This is an overview of how to use keys with lists: https://reactjs.org/docs/lists-and-keys.html

In summary, my understanding is that they're a way of making re-renders more efficient, by only updating parts of a list that have changed, rather than re-rendering the whole list. I think this is unlikely to matter on the server-side, where we only render once, but you get the warnings anyway.

Strictly speaking, you're best not using an array index as the key. As mentioned in both those doc pages, I think this is because it can confuse React if the list gets re-ordered. However, if this list is only rendered server-side and never gets re-ordered, then this probably fine.

Alternatively, if there's a way to turn off these warnings on the server-side, that might be preferable; but I've never come across one.

Just cc'ing @arelra @oliverlloyd @mxdvl to see if they have any further thoughts on this as knowledgeable React people ⚛️.

Reply:
Thanks @JamieB-gu , very well explained. I made some changes and am now using the h2 id for the key. Though, because the `TocTextElement` is recursive, the nested elements within the h2, such as sub or sup would also needed the key as well. I had to use the index of the child nodes for that: `<TocTextElement node={item} key={i.toString()} />`